File: OhHckServer.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;

public class OhHckServer extends Thread {

    public static final int PORT = 6969;
    public static final int INFO_PORT = 6970;

    private String serverName;
    private int maxPlayers;
    private int currentPlayers;
    private OhHckGame game;
    private boolean shouldContinue;
    private String hostName;
    private boolean itOpen;
    private ServerSocket serverSocket;

    public OhHckServer(String name, int maxPlayers, String host) {
        super("OhHckServer");
        this.serverName = name;
        this.hostName = host;
        this.itOpen = true;
        this.maxPlayers = maxPlayers;
        this.currentPlayers = 0;
        this.game = new OhHckGame();
        shouldContinue = true;
    }

    @Override
    public void run() {
        System.out.println("OhHckServer thread has entered run()");
        try {
            serverSocket = new ServerSocket(PORT);
            (new InfoThread()).start();
            while (shouldContinue) {
                while (currentPlayers < maxPlayers && shouldContinue) {
                    Socket sock = serverSocket.accept();
                    ServerThread st = new ServerThread(sock);
                    st.start();
                }
            }
            if (serverSocket != null) {
                serverSocket.close();
            }
        } catch (IOException e) {
            System.out.println("Exception in OhHckServer " + e.getStackTrace());
        }
        System.out.println("OhHckServer thread is returning from run()");
    }

    public class InfoThread extends Thread {

        public InfoThread() {
            super("OhHckServer$InfoThread");
        }

        @Override
        public void run() {
            System.out.println("OhHckServer$InfoThread thread has entered run()");
            try {
                ServerSocket ss = new ServerSocket(INFO_PORT);
                while (itOpen) {
                    Socket s = ss.accept();
                    PrintWriter out = new PrintWriter(s.getOutputStream(), true);
                    out.println(String.format(
                        "NAME=%s HOST=%s PLAYERS=%d/%d IPADDRESS=%s",
                        serverName, hostName, currentPlayers, maxPlayers,
                        InetAddress.getLocalHost().getHostAddress()));
                    s.close();
                    out.close();
                }
                ss.close();
            } catch (IOException e) {}
            System.out.println("OhHckServer$InfoThread thread is returning from run()");
        }
    }

    public class ServerThread extends Thread {

        private Socket socket;
        private PrintWriter out;
        private String playerName;
        private ServerSidePlayer player;

        public ServerThread(Socket socket) {
            super("OhHckServer$ServerThread-" + currentPlayers);
            this.socket = socket;
            currentPlayers++;
            this.player = new ServerSidePlayer();
            game.addClient(this);
        }

        @Override
        public void run() {
            System.out.println("OhHckServer$ServerThread thread has entered run()");
            try {
                out = new PrintWriter(socket.getOutputStream(), true);
                BufferedReader in = new BufferedReader(
                    new InputStreamReader(socket.getInputStream()));
                String input;
                game.process(null, this);
                while ((input = in.readLine()) != null && shouldContinue) {
                    game.process(input, this);
                }
                socket.close();
                out.close();
                in.close();
            } catch (IOException e) {}
            currentPlayers--;
            System.out.println("OhHckServer$ServerThread thread is returning from run()");
        }

        protected void transmit(String message) {
            if (message.contains("STOP")) {
                shouldContinue = false;
                try {
                    serverSocket.close();
                } catch (IOException e) {}
            }
            out.println(message);
        }

        public ServerSidePlayer player() {
            return player;
        }

        public int getCurrentPlayers() {
            return currentPlayers;
        }
    }

}

File: src/main/java/com/isaac/games/OhHckServer.kt
package com.isaac.games

import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.IOException
import java.io.PrintWriter
import java.net.InetAddress
import java.net.ServerSocket
import java.net.Socket

class OhHckServer(private val serverName: String, private val maxPlayers: Int, private val hostName: String) : Thread("OhHckServer") {
    private var currentPlayers: Int = 0
    private val game: OhHckGame
    private var shouldContinue: Boolean = false
    private val itOpen: Boolean
    private var serverSocket: ServerSocket? = null

    init {
        this.itOpen = true
        this.currentPlayers = 0
        this.game = OhHckGame()
        shouldContinue = true
    }

    override fun run() {
        println("OhHckServer thread has entered run()")
        try {
            serverSocket = ServerSocket(PORT)
            InfoThread().start()
            while (shouldContinue) {
                while (currentPlayers < maxPlayers && shouldContinue) {
                    val sock = serverSocket!!.accept()
                    val st = ServerThread(sock)
                    st.start()
                }
            }
            if (serverSocket != null) {
                serverSocket!!.close()
            }
        } catch (e: IOException) {
            println("Exception in OhHckServer " + e.stackTrace)
        }

        println("OhHckServer thread is returning from run()")
    }

    inner class InfoThread : Thread("OhHckServer\$InfoThread") {

        override fun run() {
            println("OhHckServer\$InfoThread thread has entered run()")
            try {
                val ss = ServerSocket(INFO_PORT)
                while (itOpen) {
                    val s = ss.accept()
                    val out = PrintWriter(s.getOutputStream(), true)
                    out.println(String.format(
                            "NAME=%s HOST=%s PLAYERS=%d/%d IPADDRESS=%s",
                            serverName, hostName, currentPlayers, maxPlayers,
                            InetAddress.getLocalHost().hostAddress))
                    s.close()
                    out.close()
                }
                ss.close()
            } catch (e: IOException) {
            }

            println("OhHckServer\$InfoThread thread is returning from run()")
        }
    }

    inner class ServerThread(private val socket: Socket) : Thread("OhHckServer\$ServerThread-$currentPlayers") {
        private var out: PrintWriter? = null
        private val playerName: String? = null
        private val player: ServerSidePlayer

        var currentPlayers: Int = 0

        init {
            currentPlayers++
            this.player = ServerSidePlayer()
            game.addClient(this)
        }

        override fun run() {
            println("OhHckServer\$ServerThread thread has entered run()")
            try {
                out = PrintWriter(socket.getOutputStream(), true)
                val `in` = BufferedReader(
                        InputStreamReader(socket.getInputStream()))
                var input: String? = `in`.readLine()
                game.process(null, this)
                while (input != null && shouldContinue) {
                    game.process(input, this)
                    input = `in`.readLine()
                }
                socket.close()
                out!!.close()
                `in`.close()
            } catch (e: IOException) {
            }

            currentPlayers--
            println("OhHckServer\$ServerThread thread is returning from run()")
        }

        fun transmit(message: String) {
            if (message.contains("STOP")) {
                shouldContinue = false
                try {
                    serverSocket!!.close()
                } catch (e: IOException) {
                }

            }
            out!!.println(message)
        }

        fun player(): ServerSidePlayer {
            return player
        }
    }

    companion object {

        val PORT = 6969
        val INFO_PORT = 6970
    }

}